aggregator/src/binary_utils.rs

```diff
@@ -572,4 +599,74 @@ mod tests {
                 .starts_with("failed to update filter:")
         );
     }
+
+    #[tokio::test]
+    async fn postgres_tls_connection() {
+        install_test_trace_subscriber();
+
+        let client = container_client();
+        // We need to be careful about providing the certificate and private key to the Postgres
+        // container. The key must have '-rw-------' permissions, and both must be readable by the
+        // postgres user, which has UID 70 inside the container at time of writing. Merely mounting
+        // a host directory in the container would be insufficient, because the files' owner UIDs
+        // will not match the postgres user's UID. Instead, we create a temporary Docker volume, run
+        // a setup container with both the volume and a host directory mounted in, copy the
+        // certificate and key into the volume, and fix up their ownership (and permissions, in
+        // case those were lost on a non-POSIX host). Then, we run a second container with the volume
+        // mounted in, and use the fixed files in the volume in database configuration.
+        let volume = Volume::new();
+        let setup_image = RunnableImage::from((
+            Postgres::with_entrypoint("/bin/bash".to_string()),
+            Vec::from([
+                "-c".to_string(),
+                concat!(
+                    "cp /etc/ssl/postgresql_host/* /etc/ssl/postgresql/ && ",
+                    "chown postgres /etc/ssl/postgresql/* && ",
+                    "chmod 600 /etc/ssl/postgresql/127.0.0.1-key.pem && ",
+                    // This satisfies the ReadyCondition.
+                    "echo 'database system is ready to accept connections' >&2",
+                )
+                .to_string(),
+            ]),
+        ))
+        .with_volume((
+            fs::canonicalize("tests/tls_files")
+                .unwrap()
+                .into_os_string()
+                .into_string()
+                .unwrap(),
+            "/etc/ssl/postgresql_host",
+        ))
+        .with_volume((volume.name(), "/etc/ssl/postgresql"));
+        let setup_container = client.run(setup_image);
+        drop(setup_container);
+
+        let image = RunnableImage::from((
+            Postgres::default(),
+            Vec::from([
+                "-c".to_string(),
+                "ssl=on".to_string(),
+                "-c".to_string(),
+                "ssl_cert_file=/etc/ssl/postgresql/127.0.0.1.pem".to_string(),
+                "-c".to_string(),
+                "ssl_key_file=/etc/ssl/postgresql/127.0.0.1-key.pem".to_string(),
+            ]),
+        ))
+        .with_volume((volume.name(), "/etc/ssl/postgresql"));
+        let db_container = client.run(image);
+        const POSTGRES_DEFAULT_PORT: u16 = 5432;
+        let port = db_container.get_host_port_ipv4(POSTGRES_DEFAULT_PORT);
+
+        let db_config = DbConfig {
+            url: format!("postgres://postgres@127.0.0.1:{port}/postgres?sslmode=require")
```

Comment:
I think we would want `sslmode=verify-full` to fully test that the trust store code is working https://www.postgresql.org/docs/current/libpq-ssl.html#LIBPQ-SSL-PROTECTION. Otherwise the client will ignore the validity of the root CA cert.

Reply:
With this client, the only options available are `disable`, `prefer`, or `require`. See https://docs.rs/tokio-postgres/latest/tokio_postgres/config/struct.Config.html. I can confirm that the `require` mode is performing full certificate verification, because it returned errors when I was passing the wrong certificate into the trust store. Effectively, I think this parameter only controls policy around `StartupMessage` versus `SSLRequest` in the wire protocol, and `tokio-postgres` entirely delegates verification to the `MakeTlsConnect` implementation. The `MakeTlsConnect` implementations only get a hostname, and no configuration. Since we're using rustls, the controlling API is `DangerousClientConfig::set_certificate_verifier()`, so we're OK.